target_chains/starknet/contracts/src/reader.cairo

```diff
@@ -0,0 +1,293 @@
+use core::option::OptionTrait;
+use core::array::ArrayTrait;
+use core::keccak::cairo_keccak;
+use core::integer::u128_byte_reverse;
+use core::fmt::{Debug, Formatter};
+use pyth::util::{UNEXPECTED_OVERFLOW, UNEXPECTED_ZERO, one_shift_left_bytes_u128};
+
+#[derive(Copy, Drop, Debug, Serde, PartialEq)]
+pub enum Error {
+    UnexpectedEndOfInput,
+}
+
+impl ErrorIntoFelt252 of Into<Error, felt252> {
+    fn into(self: Error) -> felt252 {
+        match self {
+            Error::UnexpectedEndOfInput => 'unexpected end of input',
+        }
+    }
+}
+
+/// A byte array with storage format similar to `core::ByteArray`, but
+/// suitable for reading data from it.
+#[derive(Drop, Clone, Serde)]
+pub struct ByteArray {
+    // Number of bytes stored in the last item of `self.data` (or 0 if it's empty).
+    num_last_bytes: u8,
```

Comment:
In core they use `usize` for the equivelent of `num_last_bytes`, is there any performance cost in using smaller ranged integer types? I don't see anything in the docs but curious if you know why they wouldn't also make this optimization.

Reply:
All small integers are represented as `felt252`, and `usize` is the same as `u32`, so it has all the same bound checks and conversions. So I think it shouldn't make much difference in performance, but we can do some experiments later to confirm it. I just used the smallest integer type where possible because the book [recommends to choose the type based on max value](https://book.cairo-lang.org/ch02-02-data-types.html).